### PR TITLE
fix: prevent input loss when opening the popup (#3)

### LINKED ISSUE #3: Slight lag when opening the panel sometimes
Happens irregularly, but when it does, it means your first character or two is lost, meaning you don't get the right search results.

### PR DESCRIPTION
**Pull request by charlesangus**
## Summary

Fixes #3 — *"Slight lag when opening the panel sometimes"*, where the first character or two of the user's search query were getting lost.

Four-layer fix in four commits. Layers 1–3 land the optimisation; layer 4 is post-review polish addressing the Copilot review and the `tabtabtab_anchors` use case.

### Layer 1 — defer heavy work in `TabTabTabWidget.show()` (engine)

`show()` was reloading weights from disk and re-walking Nuke's menus *before* calling `super().show()` and setting focus. While that work blocked the GUI thread, queued `KeyPress` events couldn't reach the line-edit and were lost.

Reordered so the dialog is shown and the line-edit is focused first, then `weights.load()` + `refresh_items()` + `move_selection('first')` run from a `_refresh_after_show()` method dispatched via `QTimer.singleShot(0, ...)`. Pending key events drain into the focused line-edit before the deferred work blocks the thread again.

### Layer 2 — cache the menu walk in `NukePlugin`

Even after Layer 1, the deferred refresh still ran ~15 ms of recursive menu walking + ~10 ms of colour lookups on every popup invocation, briefly flickering stale results.

- `_menu_fingerprint()` walks 2 levels deep using only `.name()` calls (no `.script()` / `.shortcut()`).
- `NukePlugin.get_items()` returns a cached items list when the fingerprint matches; full re-walk only on miss.
- `NukePlugin.get_color()` caches per `actual_class`; cache cleared whenever `get_items()` does a fresh walk.

### Layer 3 — preload the widget at plugin load

The very first Tab press in a fresh Nuke session paid ~36 ms of `TabTabTabWidget.__init__` cost (initial menu walk + initial `NodeModel` build). Added a `preload()` / `schedule_preload()` API to the engine and call it from `registerNukeAction()` via `QTimer.singleShot(0, ...)` so Nuke eagerly constructs the singleton on a deferred startup tick. First user invocation now hits the warm reuse path inside `launch()`.

A separate `_preloaded_instance` strong reference keeps the unshown widget alive — Qt's window registry only tracks shown top-level windows, so the existing `weakref.proxy` alone would dangle. The strong ref is released on first `show()` (Qt takes over) **or** on `QApplication.aboutToQuit` (whichever comes first), so the original weakref-only lifetime model that avoids the on-exit segfault from dbr/tabtabtab-nuke#4 is restored before host shutdown even if the popup is never opened.

### Layer 4 — belt-and-suspenders refresh on close

Layer 2's fingerprint is cheap but shallow: deep submenu installs (e.g. a new ToolSet several levels in) don't change the fingerprint and would stay stale indefinitely. The colour cache, similarly, can mask a default-node-colour preference edit that doesn't surface in any menu change. And for hosts where the indexed item set genuinely changes a lot during a session — `tabtabtab_anchors` indexing live nodes in the script — anything cache-shaped on the show path is suspect.

Belt-and-suspenders: `close()` schedules a `_refresh_after_close()` via `QTimer.singleShot(0, ...)` that calls `plugin.invalidate_cache()` and forces a fresh walk in the background. New `TabTabTabPlugin.invalidate_cache()` hook (no-op default in core); `NukePlugin.invalidate_cache()` overrides to drop items + fingerprint + colour caches. Any change the show-path fingerprint would have missed is picked up here, so staleness is bounded to one popup behind any mid-session change. Cost is paid entirely in the post-close background path, never on the user's hot path; a re-open while the refresh is still pending is a no-op (`isVisible()` guard).

## Timing impact

User-reported timings against the unpatched code (cold Nuke session):

| Stage | Cold | Warm |
|---|---|---|
| `__init__` | 36 ms | — (cached) |
| `show()` total | 19 ms | 46 ms |
| **Time before line-edit ready** | **~57 ms** | **~46 ms** |

After this PR:

| Stage | Cold | Warm |
|---|---|---|
| `__init__` | runs in deferred preload, off the user's hot path | — |
| `show()` blocking part | `super().show()` + `setFocus` ≈ 3 ms | ≈ 3 ms |
| **Time before line-edit ready** | **~3 ms** | **~3 ms** |

Deferred refresh still runs after focus is established; on the warm path it hits Layer 2 caches and drops to single-digit ms. Belt-and-suspenders close-refresh runs in background after each close, so the next open's deferred refresh almost always finds a warm cache.

## Test plan

- [ ] Cold Nuke launch → press Tab immediately → first character of typed query is captured (no need to wait or retype)
- [ ] Press Tab → close → press Tab again rapidly → first character still captured (cached-instance path)
- [ ] Press Tab → press Tab again immediately (no characters between) → previously-created node is invoked (`[Tab][Tab]` fast-path preserved)
- [ ] Verify search results still match expected behaviour (filtering, scoring, space-prefix modes)
- [ ] Install a new gizmo *deeply nested* (under ToolSets/User/X for example) mid-session, press Tab → close → press Tab → new node appears in the list (belt-and-suspenders catches what the fingerprint missed)
- [ ] Edit a node's default colour in prefs, press Tab → close → press Tab → swatch reflects the change
- [ ] Quit Nuke after using the popup — no segfault on exit
- [ ] Quit Nuke without ever pressing Tab — no segfault on exit (preload strong-ref is released by `aboutToQuit` handler)
- [ ] If used: verify `tabtabtab_anchors` / paste_hidden plugin still sees newly-created anchors on each Tab open (refresh-on-show ensures items added since last open are visible)

## Notes

- Existing pytest suite passes (3/3 in `tabtabtab-nuke`).
- Sister-repo sync (per the `tabtabtab-core` source-of-truth rule) is a follow-up — the upstream `c005be3` and `b152083` commits modified the vendored `tabtabtab_nuke_core.py` directly, so canonical `tabtabtab-core/tabtabtab.py` is now slightly behind. Out of scope for this PR.
- No new tests added; the project's headless test setup can't exercise the Qt event-loop reordering or Nuke menu walk that this patch targets. Comprehensive test coverage is a separate effort.
- `[Tab][Tab]` ordering and Up/Down arrow timing edge cases were both raised in review and are addressed in the relevant threads.

### CODE REVIEW

File: tabtabtab_nuke_core.py [hunk 1]
```diff
@@ -786,3 +810,43 @@ def launch(plugin, space_mode_order=None):
...
+    """
+    global _tabtabtab_instance, _preloaded_instance
+    if _tabtabtab_instance is not None:
+        return
+
+    t = TabTabTabWidget(plugin, winflags=Qt.FramelessWindowHint, space_mode_order=space_mode_order)
+    # Hold a strong reference until first show(); without this the proxy
+    # below would dangle the moment this function returns because Qt's
+    # window registry only tracks widgets that have been shown.
+    _preloaded_instance = t
```
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke_core.py:871]: Good catch — fixed in 3968664. `preload()` now connects to `QApplication.aboutToQuit` and releases both `_preloaded_instance` and `_tabtabtab_instance` before Qt teardown begins. So even on the path where the user never opens the popup, the strong reference is dropped before host shutdown, restoring the original weakref-only lifetime model in time. The widget then GCs through the normal Python/Qt teardown order rather than via a held-through-exit reference.

File: tabtabtab_nuke_core.py [hunk 2]
```diff
@@ -702,6 +702,25 @@ def show(self):
...
+        # menus can take 100-400ms; if those run synchronously here, queued
+        # KeyPress events arrive before the line-edit is ready and the user's
+        # first character or two get lost (issue #3). Deferring via
+        # singleShot(0) lets Qt drain pending input events into the now-
+        # focused line-edit before the refresh blocks the GUI thread again.
+        self.input.selectAll()
+        super(TabTabTabWidget, self).show()
+        self.input.setFocus()
+
+        QtCore.QTimer.singleShot(0, self._refresh_after_show)
```
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke_core.py:726]: You're right that the show-deferred refresh was racy against `[Tab][Tab]`. Fixed in 3968664 by moving the refresh out of `show()` entirely and into `close()` — the next-open is now served by data already refreshed at the previous close. `show()` is pure (`selectAll + super().show() + setFocus`), so a queued second Tab reaches `create()` against a stable model with no in-flight refresh that could mutate it underneath.

The freshness story is "stale at most once" — any change (menu install, weights modified by another host) shows up on the popup invocation *after* it was made, because the refresh always runs in the gap between close and the next open. Acceptable per [the discussion thread on this PR](https://github.com/charlesangus/tabtabtab-nuke/pull/4); adding nodes/menu items mid-session is rare.

File: tabtabtab_nuke_core.py [hunk 3]
```diff
@@ -712,12 +731,6 @@ def show(self):
         # Restore selection to the first item, since modelReset clears it
         self.move_selection(where="first")
```
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke_core.py:742]: Right — the deferred `move_selection('first')` clobbering an early Up/Down was a real regression. Fixed in 3968664 by removing the deferred refresh from `show()` entirely. The refresh now runs after `close()` instead, so when the popup is open there's no pending model reset / selection reset that could fire underneath the user's navigation. Arrow keys, Tab-tab, and ordinary typing all see a stable model from the moment `show()` returns.

File: tabtabtab_nuke.py [hunk 4]
```diff
@@ -27,6 +27,34 @@ def _extract_node_class_from_script(script_text):
...
+    """
+    parts = []
+    try:
+        for item in menu.items():
+            name = item.name()
+            if isinstance(item, nuke.Menu):
+                if depth > 1:
+                    parts.append(("M", name, _menu_fingerprint(item, depth - 1)))
+                else:
+                    parts.append(("M", name))
```
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke.py:53]: Agreed, the 2-level fingerprint was a real soundness gap — anything installed below depth 2 (e.g. a new ToolSet several levels in) would never appear. Fixed in 3968664 by removing the fingerprint and the items cache entirely. `get_items()` now always does a fresh recursive walk via `_find_nuke_menu_items()`. That's cheap because get_items is no longer called on the user's hot path: it runs once at widget construction (during deferred preload) and once after each popup close in the background. The reviewer's "deep submenu installs are missed" failure mode is gone.

File: tabtabtab_nuke.py [hunk 5]
```diff
@@ -103,6 +151,11 @@ def get_items(self):
                 'is_node': item['is_node'],
                 'actual_class': item['actual_class'],
             }
+        self._cached_items = all_items
+        self._cached_menu_fingerprint = fingerprint
+        # Menu set changed (or first walk) — drop color cache too in case new
+        # node classes appeared or default colors were edited.
+        self._color_cache = {}
```
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke.py:169]: Good point — tying colour invalidation to menu changes meant default-colour preference edits were masked indefinitely. Fixed in 3968664 by giving the colour cache its own explicit invalidation path: `TabTabTabPlugin.invalidate_cache()` (no-op default in core), overridden by `NukePlugin.invalidate_cache()` to clear `_color_cache`. `TabTabTabWidget._refresh_after_close()` calls it on every popup close, so a default-colour edit shows up on the next open regardless of whether menus changed.

File: tabtabtab_nuke_core.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke_core.py:726]: Update on this thread — after further investigation, I'm now arguing this isn't actually a race in practice, and 1a4ea8b reverts the show()→close() refresh move.

Tracing the flow: `create()` (`tabtabtab_nuke_core.py:752`) calls `self.input.setText(" <prev_name>")` before `close()`. That emits `textChanged` → `set_filter(" <prev_name>")`. The leading-space prefix is the **pinned-name filter mode** — it places the named item at row 0 regardless of weight scoring or item ordering. Both model state and `QListView.currentIndex` persist across close→show.

So whether the deferred refresh runs before or after the queued second Tab, row 0 is the same item and `create()` invokes the same node:

- **Refresh-before-Tab**: `refresh_items()` resets the model, `move_selection('first')` puts selection at row 0 of the freshly-filtered list → still pinned to the previous node by the leading-space filter. Tab → `create(prev_node)`. ✓
- **Tab-before-refresh**: model state from last close is intact, selection still on the previous node. Tab → `create(prev_node)`. ✓

The "stale ordering / stale menu contents" concern doesn't apply because pinned-name filtering bypasses ordering, and the only way the previous node would disappear from the menu is a mid-session deletion of that specific menu entry — extremely rare, and survivable.

Restoring the deferred-refresh-on-show is the right call because the alternative (refresh-only-on-close) breaks the `tabtabtab_anchors` / paste_hidden plugin, where indexed items genuinely do change rapidly within a session.

File: tabtabtab_nuke_core.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke_core.py:742]: Update — 1a4ea8b reverts the show()→close() refresh move, so this regression is back. Acknowledged as real but closing as unlikely in practice: the gap between super().show() and the singleShot(0) callback firing is on the order of milliseconds, and a user pressing arrow keys faster than that immediately on popup open is rare. Will revisit if anyone actually reports it; otherwise the maintainability cost of preserving selection across modelReset isn't worth the rarely-hit win.

File: tabtabtab_nuke.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke.py:53]: Update — 1a4ea8b restores the fingerprint cache (the previous "always walk fresh" version was tied to the show()→close() refresh-move that I've now reverted). The fingerprint depth blind spot is still real, but mitigated by the new belt-and-suspenders close-refresh: `TabTabTabWidget._refresh_after_close` calls `plugin.invalidate_cache()` after every popup close, which forces a fresh walk. Worst case for a deep submenu install: one popup invocation goes by with the stale list, then the close-time invalidation catches it for the next open. Cost is borne entirely in the post-close background path, never on the user's hot path.

File: tabtabtab_nuke.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke.py:169]: Update — 1a4ea8b restores the colour cache and its tied-to-fingerprint clearing, but the explicit `NukePlugin.invalidate_cache()` hook (added in 3968664) survives the partial revert. It now clears items + fingerprint + colour together, and gets called from `TabTabTabWidget._refresh_after_close` on every popup close. So a default-node-colour preference edit, which doesn't change the menu fingerprint, is still picked up — just on the popup invocation after the next close, not on the very next open. Acceptable per the "stale at most once" framing the close-refresh is built around.

File: tabtabtab_nuke_core.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke_core.py:726]: **Resolution: not a bug, won't change.** The leading-space-prefixed filter set by `create()` (`input.setText(" <prev_name>")`) pins the previously-created item to row 0 by name, bypassing weight-based ordering entirely. Whether the deferred refresh runs before or after the queued second Tab, row 0 is the same item. `[Tab][Tab]` invokes the correct node either way. Resolving.

File: tabtabtab_nuke_core.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke_core.py:742]: **Resolution: acknowledged, won't fix.** Real edge case but unlikely-in-practice — gap between `super().show()` and the `singleShot(0)` callback firing is single-digit ms; users pressing arrow keys faster than that on popup open is rare enough that the maintainability cost of preserving selection across `modelReset` isn't worth it preemptively. Will revisit if anyone reports it. Resolving.

File: tabtabtab_nuke.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke.py:53]: **Resolution: mitigated by belt-and-suspenders close-refresh in 1a4ea8b.** Fingerprint-based fast path is retained for the warm-cache benefit, and `TabTabTabWidget._refresh_after_close` calls `plugin.invalidate_cache()` after every popup close — that forces a fresh walk in the background, so a deep submenu install is picked up at most one popup behind. Cost stays in the post-close background path; never on the user's hot path. Resolving.

File: tabtabtab_nuke.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (charlesangus) [reply tabtabtab_nuke.py:169]: **Resolution: mitigated by belt-and-suspenders close-refresh in 1a4ea8b.** Colour cache stays for the per-class memoisation benefit on the hot path, but `NukePlugin.invalidate_cache()` (called from `TabTabTabWidget._refresh_after_close` after every popup close) now drops it explicitly. So a default-node-colour preference edit, which doesn't change the menu fingerprint, is still picked up — at most one popup behind any change. Resolving.
